compiler/rustc_mir_transform/src/remove_zsts.rs

```diff
@@ -36,31 +36,37 @@ struct Replacer<'a, 'tcx> {
 }
 
 /// A cheap, approximate check to avoid unnecessary `layout_of` calls.
-fn maybe_zst(ty: Ty<'_>) -> bool {
+///
+/// `Some(true)` is definitely ZST; `Some(false)` is definitely *not* ZST.
+///
+/// `None` may or may not be, and must check `layout_of` to be sure.
+fn trivially_zst<'tcx>(ty: Ty<'tcx>, tcx: TyCtxt<'tcx>) -> Option<bool> {
     match ty.kind() {
+        // definitely ZST
+        ty::FnDef(..) | ty::Never => Some(true),
+        ty::Tuple(fields) if fields.is_empty() => Some(true),
+        ty::Array(_ty, len) if let Some(0) = len.try_to_target_usize(tcx) => Some(true),
         // maybe ZST (could be more precise)
         ty::Adt(..)
         | ty::Array(..)
         | ty::Closure(..)
         | ty::CoroutineClosure(..)
         | ty::Tuple(..)
-        | ty::Alias(ty::Opaque, ..) => true,
-        // definitely ZST
-        ty::FnDef(..) | ty::Never => true,
+        | ty::Alias(ty::Opaque, ..) => None,
         // unreachable or can't be ZST
-        _ => false,
+        _ => Some(false),
```

Comment:
exhaustively match here. This branch must not have false positives given that it means "definitely *not*"